### PR TITLE
Add main navigation block

### PR DESCRIPTION
**Pull request by AshGDS**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What / Why
- Add main navigation block
- https://trello.com/c/O79e9zUQ/47-build-main-navigation
- Wasn't sure if tests were needed as there's no model for this block.

### CODE REVIEW

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (matthillco) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:4]: Small nitpick, you could use shorthand here for all the padding values, either old style with four values, or new style using `padding-block` and `padding-inline`. But happy to approve it as is if you're doing it this way for clearer readability. Same comment for margins at lines 18-19.

**Review by matthillco — COMMENTED** [review]
Looks good overall, nice job. Just one comment about whether to use shorthand properties or not.

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (AshGDS) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss:4]: Thanks Matt - yep I find being more verbose more readable, but can combine them into one line if preferred.

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (matthillco) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:27]: I'm not a fan of this format, it makes it hard to search for full classnames. Might be better to expand this out to the full class name for readability and searchability.

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (matthillco) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss:4]: No worries, happy to leave these as is.

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:25]: Nitpick: could this line move up to join the others above the media query?

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:19]: Nitpick: the spacing around these elements is inconsistent - `.app-b-main-navigation` has different padding top and bottom, but then the spacing is evened out by having top margin on the list items. Is there a reason it's like this, or could we just do

```CSS
.app-b-main-navigation {
  padding: govuk-spacing(6) 0;
}

.app-b-main-navigation__listitem {
  // no margin top
}
```

Comment (AshGDS) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss:19]: Good spot - this was actually intentional for future proofing. If the block ends up with enough links to cause overflow, the rows of links will need space between them, and we need to achieve that while still having the balance of overall top and bottom spacing, which is why I wrote it in this way.

image

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (AshGDS) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss]: Thanks Matt, I have updated it to use the full classname.

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (AshGDS) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss]: Thanks, done 👍

**Comment by AshGDS** [comment]
@andysellick @matthillco thanks both - should be ready for re-review 👍

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss:19]: THAT - is nicely done sir. Excellent. I withdraw my comment 😁

File: app/views/landing_page/blocks/_main_navigation.html.erb
Comment (andysellick) [inline app/views/landing_page/blocks/_main_navigation.html.erb:17]: Nitpick: this might be cleaner using `link_to` (https://apidock.com/rails/ActionView/Helpers/UrlHelper/link_to) something like

```
<%= link_to link["label"], link["link"], class: "govuk-link govuk-link--no-underline" %>
```

File: lib/data/landing_page_content_items/landing_page.yaml
Comment (andysellick) [inline lib/data/landing_page_content_items/landing_page.yaml:5]: Nitpick: obviously this is placeholder data, but I think `link` isn't needed for the active item?

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:23]: Sorry, I've just noticed that the designs have the mobile version as what looks like a dropdown - might be worth checking with design on how this is supposed to work (in the design the active menu item is separate from the dropdown for some reason).

Screenshot 2024-10-09 at 12 52 12

Comment (andysellick) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss:23]: Obviously this makes this block much more complicated, apologies.

Comment (AshGDS) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss:23]: @andysellick No worries - thanks for spotting this. I think it's probably easier to have two blocks here, one being the existing desktop/tablet simple menu, and then separately I'll create the more complex mobile one, and show/hide each using our CSS breakpoints. Let me know if you disagree. Will investigate and update this PR once done.

Comment (andysellick) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss:23]: I get that duplicating it might be easier but duplicating stuff makes for more stuff to maintain. If you can keep it as one thing I think that'd be better, but obviously if it's too hard for now we could duplicate and fix later.

Starting to think we should have a list of tech debt...

**Comment by AshGDS** [comment]
Thanks @andysellick - I've added a mobile menu into the existing block. I've also done the PR changes as requested. It's ready for a re review but I appreciate I've added quite a bit extra code, so no rush.

I'm also getting an error regarding the code coverage being too low now, though I'm not sure if I can write any more jasmine tests, so not sure what to do here. If you know how to resolve this, please let me know!

File: app/views/landing_page/show.html.erb
Comment (andysellick) [inline app/views/landing_page/show.html.erb:2]: Thanks!

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:60]: This element gets announced by Voiceover - ideally it wouldn't.

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:1]: The mobile menu is touching the bottom of the container, should there be some padding here?

Screenshot 2024-10-10 at 09 09 12

File: lib/data/landing_page_content_items/landing_page.yaml
Comment (andysellick) [inline lib/data/landing_page_content_items/landing_page.yaml:12]: Having child menu items is really flexible, but is that a requirement? These child items aren't displayed on desktop either - how does that work? 

(sorry, feels like you've gone the extra mile here and I'm nitpicking)

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:3]: Nitpick: extra blank line.

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:24]: Nitpick: empty line.

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:35]: Nitpick: empty line.

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [inline app/assets/stylesheets/views/_landing_page/main-navigation.scss:81]: Nitpick: empty line.

**Review by andysellick — COMMENTED** [review]
Nice improvement, just a few comments.

File: lib/data/landing_page_content_items/landing_page.yaml
Comment (AshGDS) [reply lib/data/landing_page_content_items/landing_page.yaml:12]: No worries @andysellick - The design seems to show child items on the mobile version, which don't exist on the desktop version. The implementation is quite simple, the child list elements just hide if the screen resolution is mobile. Do we need to ask someone about this functionality?

**Comment by AshGDS** [comment]
Thanks @andysellick - this should be ready for re-review. I'll try and get the coverage error resolved too.

**Comment by AshGDS** [comment]
@andysellick The requirements for this have shifted a bit due to the design review, so probably worth holding off on the review for now 👍

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (andysellick) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss:1]: This now appears to be fixed 👍

**Comment by AshGDS** [comment]
Thanks @andysellick - I've done the design refactor and pushed it as a separate commit, so it should be ready for another review.

image

image

File: app/views/landing_page/blocks/_main_navigation.html.erb
Comment (andysellick) [inline app/views/landing_page/blocks/_main_navigation.html.erb:17]: I think this bit down to around line 29 could be reduplicated a little, something like

```
          <%
            link_classes = "app-b-main-nav__link govuk-link govuk-link--no-underline"
            link_classes << " app-b-main-nav__link--active" if link["active"]
          %>
          <%= tag.a(href: link["link"], class: link_classes) do %>
            <span class="app-b-main-nav__mobile-border"></span>
            <%= link["label"] %>
          <% end %>
```

**Review by andysellick — COMMENTED** [review]
Looking good, just a few things that could be shorter.

File: app/views/landing_page/blocks/_main_navigation.html.erb
Comment (andysellick) [inline app/views/landing_page/blocks/_main_navigation.html.erb:6]: Quick one - was this supposed to be a link? I think that was raised last week.

File: app/views/landing_page/blocks/_main_navigation.html.erb
Comment (AshGDS) [reply app/views/landing_page/blocks/_main_navigation.html.erb]: Thanks, have cleaned up the code and fixed the bug, good spot 👍

File: app/assets/stylesheets/views/_landing_page/main-navigation.scss
Comment (AshGDS) [reply app/assets/stylesheets/views/_landing_page/main-navigation.scss]: Thanks, renamed the class 👍

**Comment by AshGDS** [comment]
@andysellick Thanks, it should be ready for review again (the last two commits.) I will ask the BE devs about the code coverage failures again in the tests.

**Review by andysellick — APPROVED** [review]
Code looks good to me, well done. I think there's still an outstanding question over that text being a link or not, but I think I'm happy to approve this now and we can raise another PR to make that change if we need to.

**Pull request merged**
